### PR TITLE
feat: migrate rpc signing code to alloy

### LINKED ISSUE #4742: Migrate EIP-712 and signing code in `reth-rpc` to Alloy
### :warning: Base on #4737 (`alloy`) :warning:

Use `alloy-sol-types` for EIP-712 in `EthSigner`, re-implement `ethers_core::utils::hash_message` in `reth-primitives`:

1. https://github.com/paradigmxyz/reth/blob/675c6bfc390b17d274541ef251375741f48dd3c3/crates/rpc/rpc/src/eth/signer.rs#L89-L93
2. https://github.com/paradigmxyz/reth/blob/675c6bfc390b17d274541ef251375741f48dd3c3/crates/rpc/rpc/src/eth/api/sign.rs#L22-L29

### PR DESCRIPTION
**Pull request by supernovahs**
Closes #4742

### CODE REVIEW

**Review by onbjerg — APPROVED** [review]
looks good to me, pending @DaniPopes or @mattsse

**Review by mattsse — APPROVED** [review]
lgtm, but defer to @DaniPopes on how to merge this

**Review by DaniPopes — CHANGES_REQUESTED** [review]
@supernovahs Can you rebase to `alloy` (#4737) and change the target branch to it? Otherwise LGTM.

File: crates/primitives/src/lib.rs
Comment (gakonst) [inline crates/primitives/src/lib.rs:153]: This feels like common code that we'd want to have as utilities in alloy core?

File: crates/primitives/src/lib.rs
Comment (gakonst) [inline crates/primitives/src/lib.rs:118]: where was this?

**Review by gakonst — CHANGES_REQUESTED** [review]

File: crates/primitives/src/lib.rs
Comment (DaniPopes) [inline crates/primitives/src/lib.rs:122]: Use `alloy_primitives::keccak256`

File: crates/primitives/src/lib.rs
Comment (DaniPopes) [reply crates/primitives/src/lib.rs:118]:
```suggestion
```

This was moved in the big list above

File: crates/primitives/src/lib.rs
Comment (DaniPopes) [inline crates/primitives/src/lib.rs:152]:
```suggestion
    keccak256(&eth_message)
```

File: crates/primitives/src/lib.rs
Comment (DaniPopes) [reply crates/primitives/src/lib.rs:153]: I guess so, in `alloy_primitives::utils`

File: Cargo.toml
Comment (DaniPopes) [inline Cargo.toml:115]: Can you keep these under `eth` above?

File: crates/primitives/src/lib.rs
Comment (DaniPopes) [inline crates/primitives/src/lib.rs:141]:
```suggestion
```

File: crates/primitives/src/lib.rs
Comment (DaniPopes) [reply crates/primitives/src/lib.rs]: We can merge this and upstream after, if you don't mind opening a PR to [alloy-rs/core](https://github.com/alloy-rs/core) @supernovahs after this PR?

File: Cargo.toml
Comment (DaniPopes) [reply Cargo.toml:115]: Also you'll need to patch these with git version at the bottom of the file

**Comment by supernovahs** [comment]
Don't know what happened to my pc, I am seeing ethers-core errors :/  . Trying to fix .

**Comment by supernovahs** [comment]
Thanks for the help Dan

**Comment by supernovahs** [comment]
Integration test failed with no reason ?

**Review by DaniPopes — APPROVED** [review]

**Pull request merged**
